### DISCUSSION TITLE
Suggest an Expectation or an Expectation Package!

### DISCUSSION THREAD

**Discussion by kyleaton** [great-expectations/great_expectations#4276] [Expectations]
Hello! 

This category is a place for people to begin to suggest Expectations and Expectation Packages for Great Expectations! Please feel free to create a post based on Expectations or Package ideas!

Some Relevant links: 
* [Expectation Gallery](https://greatexpectations.io/expectations) to see if your Expectation already exists.
* [How to create Custom Expectations](https://docs.greatexpectations.io/docs/guides/expectations/creating_custom_expectations/overview)
* [Contributing a Package](https://docs.greatexpectations.io/docs/contributing/contributing_package/#3-contribute-to-your-package)
* [How to contribute a Custom Expectation to Great Expectations](https://docs.greatexpectations.io/docs/guides/expectations/contributing/how_to_contribute_a_custom_expectation_to_great_expectations/)

**Comment by anikolova-vmware** [comment]
Hello Team,

Can you implement a data duplication check on an entire table or a table subset?
- records in pre-selected columns are unique (no duplicated rows)

**Reply by jedisom** [reply]
Couldn't this be accomplished with multiple [expect_column_proportion_of_unique_values_to_be_between](https://greatexpectations.io/expectations/expect_column_proportion_of_unique_values_to_be_between?filterType=Backend%20support&gotoPage=1&showFilters=false&viewType=Summary) expectations?

**Comment by guido-vries** [comment]
Foreign key expecation in sql server

**Comment by heather-dice** [comment]
1. Column row count to be between
2. Column null count to be between
3. Column not null count to be between

**Comment by METTUNR29** [comment]
Identify if the columns has SSN pattern or/ phone number pattern or/ EIN pattern or/  Shipping tracking number. each is a spserate need though.

**Reply by jedisom** [reply]
Couldn't this be completed with a [expect_column_values_to_match_regex](https://greatexpectations.io/expectations/expect_column_values_to_match_regex?filterType=Backend%20support&gotoPage=1&showFilters=false&viewType=Summary) expectation that already exists?

**Comment by jedisom** [comment]
It would be great to have a rolling year, or date, window.  For example valid values are between this year plus X and this year minus Y.  Same thing could be done for months, days, etc.

**Reply by ashankergit** [reply]
This would be great to have. I would like to check if the sources have latest data for yesterday, and this would be perfect

**Comment by kujaska** [comment]
I need this, too:
Anomaly Detection on Great Expectations 
https://github.com/great-expectations/great_expectations/discussions/7161

**Comment by rogalski** [comment]
`expect_multicolumn_values_to_be_in_set`.
Looks like simple generalization over `expect_column_pair_values_to_be_in_set`, I was quite surprised it's not available according to docs.

**Comment by pcontrerasnun** [comment]
An expectation for categorical columns to check proportion of each distinct value. For example, given column A that has distinct values 1, 2, 3. An expectation to check column has X% aprox of value 1, Y% aprox of value 2 and Z% aprox of value 3

**Comment by Athangar-tec** [comment]
It would be good to have an expectation that compares two table. This could be for both delta and full mode.

**Comment by PalBence** [comment]
It would be great if there was a `expect_column_pair_values_to_not_be_equal` function. It would be basically the 'not' version of the `expect_column_pair_values_to_be_equal` function. Or is there already a way to achieve this outcome? Thanks for the answers.

**Comment by webcoderz** [comment]
Risk of data disclosure

**Comment by gerileka** [comment]
I would be great to have something like:
- Check the standard deviation of a column for time series tables. If the last received value is X standard deviation from a given window raise alarm
- I join the foreign key from @ActivateConsulting but also in the API level

**Comment by ashankergit** [comment]
I would like to have an expectation that checks if one column is not null and the other is null. We currently have if both are null or not null

**Comment by mollysrour** [comment]
Expectation that checks for regex mappings but for a list column instead of a string column

**Comment by BPilippG** [comment]
For me and my Team it would be great to have some sort of `expect_column_values_exist_between` that allows testing if a column has at least one value within a specified interval.
The existing `expect_column_distinct_values_to_contain_set` is similar but does not always do the job because
* it requires specifying a discrete set instead of an interval, and
* it uses the computationally expensive metric `column.value_counts`, i.e. reads the complete batch.

A possible use case is to regularly test large tables in an automated pipeline whether new records are present.

Thanks & Greetings

**Comment by adaboina21** [comment]
Hi ,

Can we have expectation where we compare values between two columns ?

Example : 
Doc No","Account","Dist Type","Debit","Credit","Salesperson ID","Comm Amount"
"BF03248518","12101-900-000","Accts Receivable",3886.82,0,"",""

here we can compare Credit and Debit Amount values



Thanks,

**Comment by ria3110** [comment]
Are there any expectations available to match source data (csv file) and destination data (data in table)

**Comment by cwdjankoski** [comment]
Is there a way to check for anomalies in counts of events per day - but grouped by another categorical columns i.e.
```
+------------+------------+-------------+
|    date    | event_name | event_count |
+------------+------------+-------------+
| 2024-06-21 | event_A    |         100 |
| 2024-06-21 | event_B    |        1000 |
| 2024-06-20 | event_A    |         110 |
| 2024-06-20 | event_B    |         990 |
+------------+------------+-------------+
```

Is there a way to check and fail expectation if e.g. only the count for `(event_A)` is outlier on certain date ? 
I tried looking through the new/legacy gallery and contributed expectations but could not find anything.
Thank you in advance for any comments.

**Comment by MasemeneMatlakanaBenny** [comment]
correlation expectation would be so much better and add value

**Reply by joshua-stauffer** [reply]
hey @MasemeneMatlakanaBenny, I'm a little late to seeing this, but curious to hear more - what did you have in mind?
